### PR TITLE
VM dynamic scaling option granularity

### PR DESCRIPTION
**Pull request by harikrishna-patnala**
### Description

This PR introduces new granularity levels to configure VM dynamic scalability. Previously VM is configured to be dynamically scalable based on the template and global setting. Now we bringing this option to configure at service offering and VM level also.

VM can dynamically scale only when all flags are ON at VM level, template, service offering and global setting. If any of the flags is set to false then VM cannot be scalable. This result will be persisted in DB for each VM and will be honoured for that VM till it is updated.

We are introducing 'dynamicscalingallowed' parameter with permitted values of true or false for deployVM API and createServiceOffering API.

Following are the API parameter changes:
createServiceOffering API: 
          dynamicscalingenabled: an optional parameter of type Boolean with default value “true”.
deployVirtualMachine API: 
          dynamicscalingenabled: an optional parameter of type Boolean with default value “true”.

Following are the UI changes:
Service offering creation has ON/OFF switch for dynamic scaling enabled with default value true.

image

Deploy VM wizard has ON/OFF switch for dynamic scaling enabled with default value true in Advanced Mode section.

image

Added a documentation PR to support these changes
https://github.com/apache/cloudstack-documentation/pull/186

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Following are the test cases that I've covered as part of my testing on VMware environment.

**VM Creation:**

1. Created VM with Small service offering with all Flags ON 
            - VM1 (scalable)
2. Created VM with Small service offering which is not dynamic, with Dynamic Scaling Enabled _true_ on VM
            - VM2 (not scalable)
3. Created VM with Small service offering, with Dynamic Scaling Enabled _false_ on VM 
            - VM3 (not scalable)
4. Created VM with Small service offering which is not dynamic, with Dynamic Scaling Enabled _false_ on VM
            - VM4 (not scalable)
5. Created VM with Small service offering with all dynamic flags ON except template 
            -  VM5 (not scalable)

**VM Update:**

1. Stop VM1 => set Dynamic Scaling Enabled to _false_ => Update VM
            - VM is updated to Dynamic Scaling Enabled value _false_
2. Stop VM2 or VM3 or VM4 or VM5 => set Dynamic Scaling Enabled to _true_ => Update VM
            - VM update operation fails because VM cannot be scaled with any of the template/service offering/global setting dynamic scaling enabled flag with _false_

**VM dynamic Scaling:**

While trying to scale VM from UI, it will only list service offering that are suitable to scale and also offering with "Dynamic Scaling Enabled" value same as old service offering.

1. Create Big Service Offering with Dynamic Scaling Enabled _true_
            - SO1
2. Create Big Service Offering with Dynamic Scaling Enabled _false_
            - SO2
3. Scale VM1 (scalable) to Service Offering SO1
            - Successful
4. Scale VM1 (scalable) to Service Offering SO2
            - Operation failed because, VM1 Service Offering is marked Dynamic Scaling Enabled but not on SO2

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/ApiConstants.java [hunk 1]
```diff
@@ -831,6 +831,7 @@
     public static final String CROSS_ZONES = "crossZones";
     public static final String TEMPLATETYPE = "templatetype";
     public static final String SOURCETEMPLATEID = "sourcetemplateid";
+    public static final String DynamicScalingEnabled = "dynamicscalingenabled";
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/ApiConstants.java:834]:
```suggestion
    public static final String DYNAMIC_SCALING_ENABLED = "dynamicscalingenabled";
```

File: api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java [hunk 2]
```diff
@@ -475,4 +478,12 @@ public void setVsphereStoragePolicy(String vsphereStoragePolicy) {
     public void setRootDiskSize(Long rootDiskSize) {
         this.rootDiskSize = rootDiskSize;
     }
+
+    public Boolean getDynamicscalingenabled() {
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java:482]:
```suggestion
    public Boolean getDynamicScalingEnabled() {
```

File: api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java [hunk 3]
```diff
@@ -475,4 +478,12 @@ public void setVsphereStoragePolicy(String vsphereStoragePolicy) {
...
+        return dynamicscalingenabled;
+    }
+
+    public void setDynamicscalingenabled(Boolean dynamicscalingenabled) {
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java:486]:
```suggestion
    public void setDynamicScalingEnabled(Boolean dynamicscalingenabled) {
```

File: api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java [hunk 4]
```diff
@@ -204,6 +204,10 @@
     @Param(description = "Root disk size in GB", since = "4.15")
     private Long rootDiskSize;
 
+    @SerializedName("dynamicscalingenabled")
+    @Param(description = "true if virtual machine needs to be dynamically scalable of cpu or memory")
+    private Boolean dynamicscalingenabled;
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java:209]:
```suggestion
    private Boolean dynamicScalingEnabled;
```

File: api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java [hunk 5]
```diff
@@ -204,6 +204,10 @@
     @Param(description = "Root disk size in GB", since = "4.15")
     private Long rootDiskSize;
 
+    @SerializedName("dynamicscalingenabled")
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java:207]: Use the one defined in ApiConstants here

**Comment by rhtyd** [comment]
@blueorangutan ui

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test matrix

**Comment by harikrishna-patnala** [comment]
These failed test cases are commonly seen on all PRs, not related to this PR changes.

test_01_scale_vm on XenServer is failing with License error

**Comment by harikrishna-patnala** [comment]
@blueorangutan ui

**Review by utchoang — APPROVED** [review]
UI Code LGTM!

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 6]
```diff
@@ -3899,6 +3914,15 @@ private UserVm createVirtualMachine(DataCenter zone, ServiceOffering serviceOffe
         return vm;
     }
 
+    private Boolean checkIfDynamicScalingCanBeEnabled(Boolean dynamicScalingEnabled, ServiceOfferingVO offering, VMTemplateVO template, Long zoneId) {
```
Comment (Pearl1594) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:3917]: can we not define this in UserVMManager and call this method at all other places where the check is repeated - eg. at KubernetesClusterStartWorker and other classes

File: engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java [hunk 7]
```diff
@@ -75,6 +75,9 @@
     @Column(name = "deployment_planner")
     private String deploymentPlanner = null;
 
+    @Column(name = "dynamic_scaling_enabled")
+    private boolean dynamicScalingEnabled;
+
```
[the remark this answers is not in the document]
Comment (harikrishna-patnala) [reply engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java:80]: is_dynamic is related to custom compute offerings and thats a transient parameter.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java [hunk 8]
```diff
@@ -223,6 +223,11 @@
     @Parameter(name = ApiConstants.STORAGE_POLICY, type = CommandType.UUID, entityType = VsphereStoragePoliciesResponse.class,required = false, description = "Name of the storage policy defined at vCenter, this is applicable only for VMware", since = "4.15")
     private Long storagePolicy;
 
+    @Parameter(name = ApiConstants.DYNAMIC_SCALING_ENABLED,
+            type = CommandType.BOOLEAN,
+            description = "true if virtual machine needs to be dynamically scalable of cpu or memory")
+    protected Boolean isDynamicScalingEnabled;
+
```
[the remark this answers is not in the document]
Comment (harikrishna-patnala) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java:229]: Added since attribute

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 9]
```diff
@@ -3899,6 +3914,16 @@ private UserVm createVirtualMachine(DataCenter zone, ServiceOffering serviceOffe
         return vm;
     }
 
+    @Override
+    public Boolean checkIfDynamicScalingCanBeEnabled(ServiceOffering offering, VirtualMachineTemplate template, Long zoneId) {
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:3918]: Can this method also accept a VM as a parameter and check if the VM is scalable as well?

**Review by nvazquez — COMMENTED** [review]
Nice feature, I left some comments

File: api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java [hunk 10]
```diff
@@ -433,6 +437,10 @@ public Long getStoragePolicy() {
         return storagePolicy;
     }
 
+    public Boolean getDynamicScalingEnabled() {
+        return isDynamicScalingEnabled == null ? Boolean.TRUE : isDynamicScalingEnabled;
```
[the remark this answers is not in the document]
Comment (harikrishna-patnala) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java:441]: fixed all Boolean to boolean, thanks for that.
Yes all old service offerings prior to upgrade will be marked as dynamically scalable. This is made not to break existing VMs because service offering was not part of dynamic scaling decision before. After upgrade we need to bypass that check for old VMs, which can be achieved by marking them true.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 9 above)
Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: VM is separated because first VM deployment will not have VM flag already saved. But I felt putting a null check would simplify the code a bit. Made those changes.

**Comment by harikrishna-patnala** [comment]
@sureshanaparti @Pearl1594 @nvazquez I've addressed your comments and updated the code as per the suggestions. Please review.

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java (hunk 10 above)
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java:441]: The default value is false, for the global setting "_enable.dynamic.scale.vm_" and parameter "_isdynamicallyscalable_" in registerTemplate/ISO call , defined earlier. Any reason, to make default as "True" here ?

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 11]
```diff
@@ -623,6 +627,10 @@ public Boolean getBootIntoSetup() {
         return bootIntoSetup;
     }
 
+    public boolean isDynamicScalingEnabled() {
+        return dynamicScalingEnabled == null ? true : dynamicScalingEnabled;
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:631]: Same here as well, reason to make default as "True" ?

File: engine/schema/src/main/java/com/cloud/vm/VMInstanceVO.java [hunk 12]
```diff
@@ -57,7 +57,8 @@
 @Table(name = "vm_instance")
 @Inheritance(strategy = InheritanceType.JOINED)
 @DiscriminatorColumn(name = "type", discriminatorType = DiscriminatorType.STRING, length = 32)
-public class VMInstanceVO implements VirtualMachine, FiniteStateObject<State, VirtualMachine.Event> {
+public class
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/vm/VMInstanceVO.java:60]: extra white space here ?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 13]
```diff
@@ -2730,6 +2730,19 @@ public UserVm updateVirtualMachine(long id, String displayName, String group, Bo
 
         if (isDynamicallyScalable == null) {
             isDynamicallyScalable = vm.isDynamicallyScalable();
+        } else {
+            if (isDynamicallyScalable == true) {
+                VMTemplateVO template = _templateDao.findByIdIncludingRemoved(vm.getTemplateId());
+                if (!template.isDynamicallyScalable()) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2736]: below checks will be missed for the updated `isDynamicallyScalable` when "isDynamicallyScalable == null"

File: server/src/main/java/com/cloud/hypervisor/HypervisorGuruBase.java [hunk 14]
```diff
@@ -240,7 +239,7 @@ protected VirtualMachineTO toVirtualMachineTO(VirtualMachineProfile vmProfile) {
         // Workaround to make sure the TO has the UUID we need for Niciri integration
         VMInstanceVO vmInstance = _virtualMachineDao.findById(to.getId());
         // check if XStools/VMWare tools are present in the VM and dynamic scaling feature is enabled (per zone/global)
-        Boolean isDynamicallyScalable = vmInstance.isDynamicallyScalable() && UserVmManager.EnableDynamicallyScaleVm.valueIn(vm.getDataCenterId());
+        Boolean isDynamicallyScalable = vmInstance.isDynamicallyScalable();
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/hypervisor/HypervisorGuruBase.java:242]: where this global setting value "_UserVmManager.EnableDynamicallyScaleVm.valueIn(vm.getDataCenterId())_" is being checked now ? Please update the above comment appropriately.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 15]
```diff
@@ -3899,6 +3914,16 @@ private UserVm createVirtualMachine(DataCenter zone, ServiceOffering serviceOffe
         return vm;
     }
 
+    @Override
+    public boolean checkIfDynamicScalingCanBeEnabled(VirtualMachine vm, ServiceOffering offering, VirtualMachineTemplate template, Long zoneId) {
+        boolean canEnableDynamicScaling = (vm != null? vm.isDynamicallyScalable():true) && offering.isDynamicScalingEnabled() && template.isDynamicallyScalable() && UserVmManager.EnableDynamicallyScaleVm.valueIn(zoneId);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:3919]: format update  _vm != null ? vm.isDynamicallyScalable() : true_

File: api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java (hunk 10 above)
Comment (harikrishna-patnala) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java:441]: 1. parameter on template registration is purely a property of a template which defines whether template has hypervisor supported tools installed or not. It is not an option. So default is assumed to be "false". If tools are installed on template then admin will mark it explicitly to "true".
2. default value of global setting is also set to "false" assuming admin will enable the feature if they want overall on CloudStack management zones.
3. When coming to VM or Service offering, these are options to enable or disable at VM level. Given a template is dynamically scalable and global setting is set to true, User can still have option to decide whether he/she wants the VM to dynamically scalable or not. Defaulted to "true" because by if no option is given when deploying the VM then it adheres the template property and global setting by admin.  If we default it to "false" and User did not set any option then even though the template is dynamically scalable and global setting is "true", VM cannot scale.

In summary VM can be scalable only when all options are true. Any parameter setting to "false" will make the VM not dynamically scalable. Each parameter acts as kill switch.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 13 above)
Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2736]: This is part of updateVirtualMachine API. If no option is provided then we dont have to evaluate the existing dynamic scalability of VM. 
If user wants to update the value to "false", meaning VM not to be scalable this implies a kill switch. No need to evaluate the other values at service offering or template or global setting. CS can straight away update the VM.
Only If user wants to update the value to "true" meaning VM to be scalable then CS needs to evaluate whether VM can be marked dynamically scalable based on other options at other granular levels.

File: server/src/main/java/com/cloud/hypervisor/HypervisorGuruBase.java (hunk 14 above)
Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/hypervisor/HypervisorGuruBase.java:242]: During VM deployment when VM entry is persisted, VM's dynamic scalability is evaluated based on all options user input, template, service offering and global setting. So no need to check with global setting while preparing the transfer object.
Removing the comment.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 15 above)
Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:3919]: Addressed

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test matrix

File: api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java (hunk 10 above)
Comment (sureshanaparti) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java:441]: Ok @harikrishna-patnala thanks for the detailed explanation. Assuming the same is updated in the CS documentation (wherever required).

**Comment by sureshanaparti** [comment]
@harikrishna-patnala code LGTM, just check and confirm on the test failures above.

**Comment by harikrishna-patnala** [comment]
Thanks @sureshanaparti. 

Test failures are not related this PR changes and are common on all other PRs.
test_01_scale_vm on XenServer is failing with License issues. So I've manually ran it on my setup to check the actual result and it passed.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (nvazquez) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/offering/CreateServiceOfferingCmd.java:441]: Copy, thanks

**Review by nvazquez — APPROVED** [review]
LGTM

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@harikrishna-patnala can you address the conflicts, please?

**Comment by harikrishna-patnala** [comment]
> @harikrishna-patnala can you address the conflicts, please?

Resolved conflicts.

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Review by DaanHoogland — COMMENTED** [review]
a lot of translations know contain english texts, shouldn't these just be left out?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 16]
```diff
@@ -7045,13 +7072,18 @@ public UserVm restoreVirtualMachine(final Account caller, final long vmId, final
...
                         vm.setTemplateId(newTemplateId);
-                        _vmDao.update(vmId, vm);
                     }
+                    // check if VM can be dynamically scalable with the new template
+                    ServiceOfferingVO serviceOffering = _offeringDao.findById(vm.getServiceOfferingId());
+                    VMTemplateVO newTemplate = _templateDao.findById(newTemplateId);
+                    boolean dynamicScalingEnabled = checkIfDynamicScalingCanBeEnabled(vm, serviceOffering, newTemplate, vm.getDataCenterId());
+                    vm.setDynamicallyScalable(dynamicScalingEnabled);
+                    _vmDao.update(vmId, vm);
+
```
[the remark this answers is not in the document]
Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: Done

File: api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java [hunk 17]
```diff
@@ -204,6 +204,10 @@
     @Param(description = "Root disk size in GB", since = "4.15")
     private Long rootDiskSize;
 
+    @SerializedName(ApiConstants.DYNAMIC_SCALING_ENABLED)
+    @Param(description = "true if virtual machine needs to be dynamically scalable of cpu or memory")
```
[the remark this answers is not in the document]
Comment (harikrishna-patnala) [reply api/src/main/java/org/apache/cloudstack/api/response/ServiceOfferingResponse.java]: updated based on review comments

File: engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java [hunk 18]
```diff
@@ -105,64 +108,22 @@ public ServiceOfferingVO(String name, Integer cpu, Integer ramSize, Integer spee
...
+                             VirtualMachine.Type vmType, String hostTag, String deploymentPlanner, Boolean dynamicScalingEnabled) {
         super(name, displayText, provisioningType, false, tags, recreatable, useLocalStorage, systemUse, true);
         this.cpu = cpu;
         this.ramSize = ramSize;
         this.speed = speed;
         this.rateMbps = rateMbps;
         this.multicastRateMbps = multicastRateMbps;
         this.offerHA = offerHA;
-        this.limitCpuUse = limitCpuUse;
+        this.limitCpuUse = limitResourceUse;
```
[the remark this answers is not in the document]
Comment (harikrishna-patnala) [reply engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java:121]: There are some unused constructors, so this is the result of removing/clubbing them.

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
I will retest the xen and kvm environments, but I don't consider these errors above to be due to this PR. tested functionality as described. LGTM

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 kvm-centos7

**Review by DaanHoogland — APPROVED** [review]
LGTM (tested)

**Comment by DaanHoogland** [comment]
the scaling tests in Travis are due to the template not being marked scalable in the test env. This error seems relted but is happening on other PRs as well so I don't consider it a blocker for this PR.

**Comment by harikrishna-patnala** [comment]
There is a test failure in the simulator environment, I've fixed it in the last commit. All checks for good now.

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by alexandremattioli** [comment]
LGTM

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@blueorangutan test centos7 xenserver-71
